memory/src/main/scala/filodb.memory/data/OffheapSortedIDMap.scala

```diff
@@ -0,0 +1,633 @@
+    mapWithShared(binarySearch(key) >= 0)
+  }
+
+  /**
+   * Returns the first element, the one with the lowest key. Caller must hold any lock.
+   * Throws IndexOutOfBoundsException if there are no elements.
+   */
+  final def mapGetFirst(): NativePointer = {
+    if (size <= 0) {
+      throw new IndexOutOfBoundsException
+    }
+    arrayGet(first)
+  }
+
+  /**
+   * Returns the last element, the one with the highest key. Caller must hold any lock.
+   * Throws IndexOutOfBoundsException if there are no elements.
+   */
+  final def mapGetLast(): NativePointer = {
+    if (size <= 0) {
+      throw new IndexOutOfBoundsException
+    }
+    arrayGet(realIndex(first + size - 1))
+  }
+
+  /**
+   * Produces an ElementIterator for going through every element of the map in increasing key order.
+   */
+  final def mapIterate(): ElementIterator = {
+    new LazyElementIterator(() => {
+      mapAcquireShared()
+      try {
+        new MapIterator(first, first + size)
+      } catch {
+        case e: Throwable => mapReleaseShared(); throw e;
+      }
+    })
+  }
+
+  /**
+   * Produces an ElementIterator for iterating elements in increasing key order from startKey
+   * to endKey.
+   * @param startKey start at element whose key is equal or immediately greater than startKey
+   * @param endKey end iteration when element is greater than endKey.  endKey is inclusive.
+   */
+  final def mapSlice(startKey: Long, endKey: Long): ElementIterator = {
+    new LazyElementIterator(() => {
+      mapAcquireShared()
+      try {
+        new MapIterator(binarySearch(startKey) & 0x7fffffff, first + size) {
+          override def isPastEnd: Boolean = mapKeyRetrieve(getNextElem) > endKey
+        }
+      } catch {
+        case e: Throwable => mapReleaseShared(); throw e;
+      }
+    })
+  }
+
+  /**
+   * Produces an ElementIterator for iterating elements in increasing key order starting
+   * with startKey.
+   * @param startKey start at element whose key is equal or immediately greater than startKey
+   */
+  final def mapSliceToEnd(startKey: Long): ElementIterator = {
+    new LazyElementIterator(() => {
+      mapAcquireShared()
+      try {
+        new MapIterator(binarySearch(startKey) & 0x7fffffff, first + size)
+      } catch {
+        case e: Throwable => mapReleaseShared(); throw e;
+      }
+    })
+  }
+
+  /**
+   * Acquire exclusive access to this map, spinning if necessary. Exclusive lock isn't re-entrant.
+   */
+  final def mapAcquireExclusive(): Unit = {
+    // Spin-lock implementation. Because the owner of the shared lock might be blocked by this
+    // thread as it waits for an exclusive lock, deadlock is possible. To mitigate this problem,
+    // timeout and retry, allowing shared lock waiters to make progress. The timeout doubles
+    // for each retry, up to a limit, but the retries continue indefinitely.
+
+    var timeoutNanos = InitialExclusiveRetryTimeoutNanos
+    var warned = false
+
+    while (true) {
```

Comment:
Possibility of constant spinning. There be a timeout from this method by returning a false.
We should explore how the caller should handle the situation when lock could not be obtained.

Reply:
ah I see there is a tryAcquireExclusive. Should we even have this method then ?

Reply:
This was the same code in OffheapLFSortedIdMap. The internal method is a bit clearer methinks.